File: lunatica-be/app/src/main/java/uk/matvey/lunatica/account/AccountRepo.java
package uk.matvey.lunatica.account;

import com.google.cloud.firestore.Firestore;
import org.springframework.stereotype.Component;

import java.util.HashMap;
import java.util.Map;

@Component
public class AccountRepo {
    private final Firestore db;

    public AccountRepo(Firestore db) {
        this.db = db;
    }

    public void add(Account account) {
        db.collection("accounts").document(account.id.toString()).set(toDoc(account));
    }

    private static Map<String, Object> toDoc(Account account) {
        HashMap<String, Object> doc = new HashMap<>();
        doc.put("email", account.email);
        account.tgChatId.ifPresent(tgChatId -> doc.put("tgChatId", tgChatId));
        return doc;
    }
}

File: lunatica-be/app/src/main/kotlin/uk/matvey/lunatica/account/AccountRepo.kt
package uk.matvey.lunatica.account

import com.google.cloud.firestore.Firestore
import org.springframework.stereotype.Component
import java.time.Instant
import java.util.UUID

@Component
class AccountRepo(private val db: Firestore) {
    fun add(account: Account) {
        db.collection("accounts").document(account.id.toString()).set(account.toDoc())
    }

    fun get(id: UUID): Account {
        return db.collection("accounts").document(id.toString()).get().get()
            .let { fromDoc(id.toString(), requireNotNull(it.data)) }
    }

    fun update(account: Account) {
        db.collection("accounts").document(account.id.toString())
            .set(account.toDoc())
    }

    fun getByTgChatId(tgChatId: Long): Account {
        return db.collection("accounts").whereEqualTo("tgChatId", tgChatId).get().get().single()
            .let { fromDoc(it.id, requireNotNull(it.data)) }
    }

    companion object {
        private fun Account.toDoc(): Map<String, Any> {
            return listOfNotNull(
                "email" to this.email,
                "passHash" to this.passHash,
                "createdAt" to this.createdAt.toString(),
                "updatedAt" to this.updatedAt.toString(),
                this.tgChatId?.let { "tgChatId" to it },
            ).toMap()
        }

        private fun fromDoc(id: String, doc: Map<String, Any>): Account {
            return Account(
                UUID.fromString(id),
                doc.getValue("email").toString(),
                doc.getValue("passHash").toString(),
                doc["tgChatId"] as? Long,
                Instant.parse(doc.getValue("createdAt").toString()),
                Instant.parse(doc.getValue("updatedAt").toString())
            )
        }
    }
}
